includes/class-fontawesome.php

```diff
@@ -1087,6 +1212,110 @@ public function common_data_for_js_bundle() {
 		);
 	}
 
+	/**
+	 * Enqueues a kit loader <script>.
+	 *
+	 * A proper kit loader <script>  looks like this:
+	 *
+	 * <script src="https://kit.fontawesome.com/deadbeef00.js" crossorigin="anonymous"></script>
+	 * 
+	 * where deadbeef00 is the kitToken
+	 *
+	 * Internal use only. Not part of this plugin's public API.
+	 *
+	 * @ignore
+	 * @internal
+	 * @throws FontAwesome_ConfigurationException if the kit_token is not a string
+	 */
+	public function enqueue_kit( $kit_token ) {
+		if ( ! is_string( $kit_token ) ) {
+			throw new FontAwesome_ConfigurationException(
+				'Font Awesome is configured to use a kit, but no valid kit token was provided.'
+			);
+		}
+
+		add_action(
+			'wp_enqueue_scripts',
+			function () use( $kit_token ) {
+				wp_enqueue_script(
+					self::RESOURCE_HANDLE,
+					trailingslashit( self::KIT_LOADER_BASE_URL ) . $kit_token . '.js',
+					[],
+					null,
+					false
+				);
+
+				/**
+				 * Kits have built-in support for detecting conflicts, but we need to
+				 * inject some configuration to turn it on. We will do that by manipulating
+				 * the FontAwesomeKitConfig global property.
+				 */
+				if ( $this->detecting_conflicts() ) {
+					/**
+					 * Kits Conflict Detection expects this value to be in milliseconds
+					 * since the unix epoch.
+					 */
+					$detect_conflicts_until = $this->options()['detectConflictsUntil'] * 1000;
+
+					wp_add_inline_script(
```

Comment:
@robmadole I'm not sure whether to think of this as elegance or hackery, but I want to highlight it and request code review specifically here.

One of the challenges I had anticipated a week or so ago was how to deal with the possibility that conflict detection might be enabled both by the WordPress plugin _and_ within a kit loaded by the plugin, once we added kits support. It would probably be chaos if two conflict detectors were loaded and fighting one another, possibly getting into infinite recursion or some such.

So the maybe-elegant part of this solution is a sort of backwards overriding of the `FontAwesomeKitConfig`. When a kit is loaded in WordPress, we need to be able to override its `detectConflictsUntil` config--the config that is normally prepended to the body of the kit loader script by the edge app.

But that config is a global property. So if we guarantee that a script loaded by WordPress _before_ the kit loader script sets up that global property _first_, with a setter to control future settings, then we can switch the order of priority in that sitter so that the config we inject early (by the WordPress plugin), is used to override whatever settings are added by the kit loader later. Maybe that's the hacky part.

It works. And I tried this in several different browsers, since defining JS Object properties with getters and setters is not a tool I've reached for much, if ever. And it works everywhere I tried it.

So, curious about your thoughts here:
-  is this elegant or gross?
- is this a temporary hack, to be replaced by something better eventually, or a fine long-term solution?

Reply:
Ok, just looked over the code and I think I understand what it's doing. Let me start by answering the questions:

> elegant or gross?

I think it's probably elegant. Gross is not how I would describe it but it's possibly brittle. That's my concern here I think.

> is this a temporary hack, to be replaced by something better eventually, or a fine long-term solution?

I think if you can assuage my feeling of brittle-ness it's fine long-term. It's monkey-patching no doubt but there are reasons to monkey-patch. What we can do from a Kit perspective is define a loose contract of how we configure it. By that I mean we agree: "the kit JS will always start by assigning window.FontAwesomeKitConfig equal to some plain JS object". That contract needs to always be in place for this to work. (If down the road we chose to configure kits differently we'd break the WP plugin.)

So let me go the other direction with this and see what you think. If you tell me why this _won't_ work I think your explanation may satiate my anxiety.

What I would have tried first is: 1.) let the Kit load _first_ so `window.FontAwesomeKitConfig` is already defined, 2.) have WP inline script load _after_ and simply overwrite it.

```
window.__FontAwesome__WP__KitConfig__ = {
 	detectConflictsUntil: $detect_conflicts_until
 }
window.FontAwesomeKitConfig.detectConflictsUntil = $detect_conflicts_until
```

What I don't understand about what you did with this getter/setter thing is how it continues to work when the kit does this: `window.FontAwesomeKitConfig = { ... }`. Doesn't that replace the object in memory? There is some sorcery there I'm not grokking.

Reply:
The problem I'm trying to resolve is that, once the kit loader reads its own config (prepended by the edge app), it makes its decision about loading the conflict detector and goes ahead and loads it (or not), along with other decisions it makes based on the value of `detectConflictsUntil`, like adding the custom data attributes to _ignore_ known non-conflicts (crucial for proper functioning).

It would simply be too late if the WordPress plugin's inline script tried to override that `detectConflictsUntil` value after the kit loader script has already loaded and (synchronously) run.

So it's precisely this assignment that happens (first thing) in the kit loader that I'm trying monkey-patch:
`window.FontAwesomeKitConfig = { ... }`

Because the WordPress inline script is loaded _before_ that assignment, and will have already created `FontAwesomeKitConfig` as a property with a setter, then by the time the kit loader does that initial assignment, the assignment operation invokes the _setter_. And that's the sorcery that lets this whole thing work: but turning that "assignment" in a setter invocation. In the setter, we can define our own logic for who overrides whom.

Reply:
OHH. I see now. Ok, I fully understand. Not monkey-patching, more like gorilla-hijacking.

I don't think this is brittle then as long as it works that way in all the target browsers (and sounds like you've already tested that.)

I'm good with this :+1:

As traditional though: **one more thought**.

What if you could load the Kit and pass it the detect conflicts value like this? Would this change the game?

`https://kit.fontawesome.com/abcdef01234.js?detect-conflicts=1239392332`

Reply:
Yes, that would change the game. But also if I were going to parameterize this one on the kit, I think I'd first reach for a custom data attribute on the HTML. In fact, I seem to recall _almost_ adding that when I added conflict detection to the kit loader.

So it would be like:

`<script src="https://kit.fontawesome.com/abcdef01234.js" data-detect-conflicts-until="1239392332"></script>`

That `data-` attribute would override the value in `FontAwesomeKitConfig.detectConflictsUntil`.

I almost reached for that change as my first choice in this case. And I do think it would be cleaner. But I was unsure at the moment about whether this would be backwards compatible with other kit loader scripts already out there. It would really need to work immediately for all live kits. When I was sure that the gorilla-hijacking was going to work for all cases, and I hadn't yet reasoned through whether parameterizing the kit loader in the current edge app would work for all cases, I opted for the former.

But if you want to take this the other direction, I'm good with that (though concerned about the time and effort required, given that we're trying to be in "only tying up loose ends" mode here).

Reply:
Nah, stick with your current implementation, Mike. (For level of effort reasons.)

We don't have the new kits infrastructure deployed yet so it's a moot point anyway. I was just thinking out loud more than anything.